File: app/src/main/java/com/baikas/sporthub6/models/chat/ChatMessageBatch.java
package com.baikas.sporthub6.models.chat;


import com.baikas.sporthub6.helpers.time.TimeFromInternet;

import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ChatMessageBatch {

    @NotNull
    private String id;
    @NotNull
    private String chatId;
    public static final int MAX_MESSAGES = 20;

    private List<ChatMessage> chatMessages = new ArrayList<>();
    private List<String> seenByUsersId = new ArrayList<>();
    private List<String> usersWithActivity = new ArrayList<>();
    private long createdAtUTC;
    private long lastModifiedTimeInUTC;


    public ChatMessageBatch(@NotNull String id, @NotNull String chatId,List<ChatMessage> chatMessages,List<String> seenByUsersId,List<String> usersWithActivity, long createdAtUTC) {
        this.id = id;
        this.chatId = chatId;
        this.chatMessages = chatMessages;
        this.seenByUsersId = seenByUsersId;
        this.usersWithActivity = usersWithActivity;
        this.createdAtUTC = createdAtUTC;
        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();
    }

    public ChatMessageBatch(@NotNull Map<String, Object> map) {
        this.id = (String) map.get("id");
        this.chatId = (String) map.get("chatId");

        List<Map<String, Object>> chatMessagesList = (List<Map<String, Object>>) map.get("chatMessages");

        this.chatMessages = chatMessagesList.stream()
                .map((chatMessageMap) -> new ChatMessage(chatMessageMap))
                .collect(Collectors.toList());

        this.seenByUsersId = (List<String>)map.get("seenByUsersId");
        this.usersWithActivity = (List<String>)map.get("usersWithActivity");

        if (map.get("createdAtUTC") instanceof Integer){
            this.createdAtUTC = ((Integer) map.get("createdAtUTC")).longValue();
        }else
            this.createdAtUTC = (Long) map.get("createdAtUTC");


        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();
    }


    // Copy constructor
    public ChatMessageBatch(ChatMessageBatch other) {
        this.id = other.id;
        this.chatId = other.chatId;
        // Deep copy for list to ensure the lists are independent
        this.chatMessages = new ArrayList<>(other.chatMessages);
        this.usersWithActivity = new ArrayList<>(other.usersWithActivity);
        this.seenByUsersId = new ArrayList<>(other.seenByUsersId);
        this.createdAtUTC = other.createdAtUTC;
        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();
    }

    @NotNull
    public String getId() {
        return id;
    }

    public void setId(@NotNull String id) {
        this.id = id;
    }

    @NotNull
    public String getChatId() {
        return chatId;
    }

    public void setChatId(@NotNull String chatId) {
        this.chatId = chatId;
    }

    public List<ChatMessage> getChatMessages() {
        return chatMessages;
    }

    public void setChatMessages(List<ChatMessage> chatMessages) {
        this.chatMessages = chatMessages;
    }


    public long getCreatedAtUTC() {
        return createdAtUTC;
    }

    public void setCreatedAtUTC(long createdAtUTC) {
        this.createdAtUTC = createdAtUTC;
    }

    public long getLastModifiedTimeInUTC() {
        return lastModifiedTimeInUTC;
    }

    public List<String> getSeenByUsersId() {
        return seenByUsersId;
    }

    public void setSeenByUsersId(List<String> seenByUsersId) {
        this.seenByUsersId = seenByUsersId;
    }

    public List<String> getUsersWithActivity() {
        return usersWithActivity;
    }

    public void setUsersWithActivity(List<String> usersWithActivity) {
        this.usersWithActivity = usersWithActivity;
    }
}

File: functions/src/models/chat/ChatMessageBatch.ts
/* eslint-disable */


import {ChatMessage} from "./ChatMessage";
import {BadModelFieldException} from "../../exceptions/BadModelFieldException";

export class ChatMessageBatch {

    private id!:string;
    private chatId!:string;
    private chatMessages!: ChatMessage[];
    private seenByUsersId!: string[];
    private usersWithActivity!: string[];
    private createdAtUTC!:number;
    private lastModifiedTimeInUTC!:number;

    public static readonly MAX_MESSAGES:number = 20;


    constructor() {
        this.setLastModifiedTimeInUTCEmpty();
    }


    static makeChatMessageBatch(chatMessageBatchId: string, chatId: string, chatMessages: ChatMessage[], seenByUsersId: string[], usersWithActivity: string[], createdAtUTC: number) {

        const chatMessageBatch:ChatMessageBatch = new ChatMessageBatch();

        chatMessageBatch.setId(chatMessageBatchId);
        chatMessageBatch.setChatId(chatId);
        chatMessageBatch.setChatMessages(chatMessages);
        chatMessageBatch.setSeenByUsersId(seenByUsersId);
        chatMessageBatch.setUsersWithActivity(usersWithActivity);
        chatMessageBatch.setCreatedAtUTC(createdAtUTC);


        return chatMessageBatch;
    }

    static makeInstanceFromSnapshot(snapshotNeedsParsing: any): ChatMessageBatch {

        const chatMessageBatch:ChatMessageBatch = new ChatMessageBatch();

        let parsedSnapshot;
        try{
            parsedSnapshot = JSON.parse(snapshotNeedsParsing);
        }catch (Error){
            parsedSnapshot = snapshotNeedsParsing;
        }

        chatMessageBatch.setId(parsedSnapshot.id);
        chatMessageBatch.setChatId(parsedSnapshot.chatId);


        const chatMessages: ChatMessage[] = [];
        for (const chatMessage of parsedSnapshot.chatMessages) {
            chatMessages.push(ChatMessage.makeInstanceFromSnapshot(chatMessage));
        }
        chatMessageBatch.setChatMessages(chatMessages);


        chatMessageBatch.setSeenByUsersId(parsedSnapshot.seenByUsersId);
        chatMessageBatch.setUsersWithActivity(parsedSnapshot.usersWithActivity);
        chatMessageBatch.setCreatedAtUTC(parsedSnapshot.createdAtUTC);
        chatMessageBatch.setLastModifiedTimeInUTC(parsedSnapshot.lastModifiedTimeInUTC);

        return chatMessageBatch;
    }

    toObject() {//this is the output that goes to firestore so becareful with the changes in this class


        const chatMessagesForFirebase: {}[] = [];
        this.chatMessages.forEach((value: ChatMessage, key) => {
            chatMessagesForFirebase.push(value.toObject());
        });


        return {
            id: this.id,
            chatId: this.chatId,
            chatMessages: chatMessagesForFirebase,
            seenByUsersId: this.seenByUsersId,
            usersWithActivity: this.usersWithActivity,
            createdAtUTC: this.createdAtUTC,
            lastModifiedTimeInUTC: this.lastModifiedTimeInUTC,
            MAX_MESSAGES: ChatMessageBatch.MAX_MESSAGES,
        };
    }


    public getId(): string {
        return this.id;
    }

    public setId(id: string) {

        if (id === undefined || typeof id !== 'string') {
            throw new BadModelFieldException("id must be a string");
        }

        let idLength:number = 60;
        if (id.length > idLength){
            throw new BadModelFieldException("id must be less than " + idLength + " letters");
        }


        this.id = id;
    }

    public getChatId(): string {
        return this.chatId;
    }

    public setChatId(chatId: string) {

        if (chatId === undefined || typeof chatId !== 'string') {
            throw new BadModelFieldException("chat message batch id must be a string");
        }

        let maxLength:number = 60;
        if (chatId.length > maxLength){
            throw new BadModelFieldException("chat message batch id must be less than " + maxLength + " letters");
        }

        this.chatId = chatId;
    }

    public getChatMessages(): ChatMessage[] {
        return this.chatMessages;
    }

    public setChatMessages(chatMessagesList: ChatMessage[]) {

        for (const chatMessage of chatMessagesList) {

            if (chatMessage === undefined || !(chatMessage instanceof ChatMessage)) {
                throw new BadModelFieldException("something is wrong with chatMessage");
            }

        }

        this.chatMessages = chatMessagesList;
    }

    public getSeenByUsersId(): string[] {
        return this.seenByUsersId;
    }

    public setSeenByUsersId(seenByList: string[]) {

        for (const userId of seenByList) {
            if (userId === undefined || typeof userId !== 'string') {
                throw new BadModelFieldException("seenBy user id must be a valid string");
            }

            let maxLength:number = 60;
            if (userId.length > maxLength){
                throw new BadModelFieldException("seenBy user id must be less than " + maxLength + " letters");
            }
        }

        let deepCopy = JSON.parse(JSON.stringify(seenByList));


        this.seenByUsersId = deepCopy;
    }

    public getUsersWithActivity(): string[] {
        return this.usersWithActivity;
    }

    public setUsersWithActivity(usersWithActivity: string[]) {

        for (const userId of usersWithActivity) {
            if (userId === undefined || typeof userId !== 'string') {
                throw new BadModelFieldException("usersWithActivity user id must be a valid string");
            }

            let maxLength:number = 60;
            if (userId.length > maxLength){
                throw new BadModelFieldException("usersWithActivity user id must be less than " + maxLength + " letters");
            }
        }

        let deepCopy = JSON.parse(JSON.stringify(usersWithActivity));

        this.usersWithActivity = deepCopy;
    }

    public getCreatedAtUTC(): number {
        return this.createdAtUTC;
    }

    public setCreatedAtUTC(createdAtUTC: number) {

        if (createdAtUTC === undefined || typeof createdAtUTC !== 'number' || isNaN(createdAtUTC)) {
            throw new BadModelFieldException("created at must be a number");
        }


        this.createdAtUTC = createdAtUTC;
    }

    public getLastModifiedTimeInUTC(): number {
        return this.lastModifiedTimeInUTC;
    }

    public setLastModifiedTimeInUTC(lastModifiedTimeInUTC: number) {

        const currentDate = new Date();

        this.lastModifiedTimeInUTC = currentDate.getTime();
    }

    public setLastModifiedTimeInUTCEmpty() {

        const currentDate = new Date();

        this.lastModifiedTimeInUTC = currentDate.getTime();
    }


}
